src/main/scala/beam/sim/population/PopulationScaling.scala

```diff
@@ -38,90 +39,72 @@ class PopulationScaling extends LazyLogging {
     val repetitions = math.floor(additionalPopulationRequired / scenario.getPopulation.getPersons.size()).toInt
     // A counter that tracks the number of population in the current scenario (stop as soon as this counter hits the required total population)
     var populationCounter = scenario.getPopulation.getPersons.size()
+    val existingHouseHolds = scenario.getHouseholds.getHouseholds.asScala.toSeq
     for (i <- 0 to repetitions) {
-      val existingHouseHolds = scenario.getHouseholds.getHouseholds.asScala
-      val newHouseHolds: mutable.Map[HouseholdImpl, Household] = existingHouseHolds.flatMap(hh => {
-        // proceed only if the required population is not yet reached
-        if (populationCounter <= totalPopulationRequired) {
-          // get the members in the current house hold and duplicate them with new ids
-          val members = hh._2.getMemberIds.asScala
-            .flatMap(m => {
-              val person: Person = scenario.getPopulation.getPersons.get(m)
-              if (person != null) {
-                populationCounter += 1
+      // generate new households
+      existingHouseHolds.toStream
+        .takeWhile(_ => populationCounter <= totalPopulationRequired)
+        .map {
+          case (houseHoldId, houseHold) =>
+            // proceed only if the required population is not yet reached
+            // get the members in the current house hold and duplicate them with new ids
+            val members =
+              houseHold.getMemberIds.asScala
+                .flatMap(m => Option(scenario.getPopulation.getPersons.get(m)))
                 // proceed only if the required population is not yet reached
-                if (populationCounter <= totalPopulationRequired) {
+                .take((totalPopulationRequired - populationCounter + 1).toInt)
```

Comment:
This isn't the same logic - it will result in 1 more than the original code. For example, the original code:

```
totalPopReq = 4
popCtr = 1 //Starting size
```

Now, when it first hits it increments (2) and passes the if (2<=4), so run = 1. Then on the next run it increments (3) and passes the if (3<=4), so run=2. Then on the next run it increments (4) and passes the if (4<=4), so run=3. Then on the next run it increments (5) and fails the if (5<=4), so run stops at 3.

So, 3 runs with the above starting point. 

In the new code we would expect the take to be 3, but it is 4 (4-1+1=4). 

Thus the +1 should be dropped, unless you say the original code was the bug?

Reply:
Indeed, thank you, Justin. 
Missed this stuff last night :D
Fixed, also compared the results with the old ones again, also looks like old code was generating some households without people.